nemo_aligner/data/nlp/datasets.py

```diff
@@ -958,3 +962,79 @@ def collate_fn(self, batch):
+# and violating this limit causes a crash. Therefore, to avoid building tedious preprocessing scripts for every kind of data we use
+# it's easier to have a wrapper class which removes those outlier samples which are too long at runtime, which is what this class does.
+#
+# This class supports truncating by multiple fields in the data, for example:
+#     truncation_field: "context_ids,input_ids"
+#     max_seq_length: 1600, 1900
+#
+#     This means that samples need to have context_ids <= 1600 AND input_ids <= 1900 in order to be included for training
+#
+# You can also truncating by just a single field:
+#
+#     truncation_field: "context_ids"
+#     max_seq_length: 1600
+#
+class TruncatedGPTSFTChatDataset(GPTSFTChatDataset):
+    def _build_samples_mapping(self):
+        super()._build_samples_mapping()
+
+        assert self.max_seq_length is not None, "max_seq_length cannot be None if using TruncatedGPTSFTChatDataset"
+        assert self.max_num_samples is None, "max_num_samples must be None when using TruncatedGPTSFTChatDataset"
+
+        N = len(self)
+        good_idxes = []
+        for idx in range(N):
+            sample = self[idx]
+
+            if isinstance(self.max_seq_length, list):
+                assert len(self.truncation_fields) == len(
+                    self.max_seq_length
+                ), "If passing a list for `max_seq_length` it must be equal in length to `truncation_fields`"
+                if all(
+                    [len(sample[trunc]) <= limit for trunc, limit in zip(self.truncation_fields, self.max_seq_length)]
+                ):
+                    good_idxes.append(idx)
+            else:
+                if all([len(sample[trunc]) <= self.max_seq_length for trunc in self.truncation_fields]):
+                    good_idxes.append(idx)
+
+        if hasattr(self.indexed_dataset, "select"):
+            self.indexed_dataset = self.indexed_dataset.select(good_idxes)
+        else:
+            self.indexed_dataset = [x for i, x in enumerate(self.indexed_dataset) if i in good_idxes]
+
+        if isinstance(self.max_seq_length, list):
+            self.max_seq_length = max(self.max_seq_length)
+
+        logging.info(
+            f"TruncatedSFTChatDataset has {len(good_idxes)} total samples. Dropped {N - len(good_idxes)} samples."
+        )
+
+    # if we are unable to use datasets.load_dataset due to certain dataset formats causing issues, then we need to fallback
+    # to loading a jsonl manually. In that case, we need to ensure each sample sent to _process_example is immutable or else
+    # you will have corrupted text, see here: https://github.com/NVIDIA/NeMo/blob/25b7d3d7f217a11f85ed23b5917b7b840f330e2f/nemo/collections/nlp/data/language_modeling/megatron/gpt_sft_chat_dataset.py#L204
+    def _process_example(self, example):
+        if hasattr(self.indexed_dataset, "select"):
```

Comment:
Just taking a note to revisit this `if` depending on the resolution of my other comments in this file.

Reply:
Need to address https://github.com/NVIDIA/NeMo-Aligner/pull/321/files#r1945138081 first